### PR TITLE
DataTable deserialization improvements

### PR DESCRIPTION
**Pull request by richardstartin**
Calling `Enum.values()` allocates an array every time it is called, and this method gets called a lot during `DataTable` deserialization, which shows up in allocation profiles.

Also deserializes table directly from `ByteBuffer` to avoid allocating lots of `byte[]` during deserialization. This isn't done when the `DataTable` retains a copy of the `ByteBuffer` because of ambiguity of ownership, though it looks like it could be done safely.

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java [hunk 1]
```diff
@@ -350,32 +340,29 @@ private byte[] serializeMetadata()
    * This is to make V3 implementation keep the consumers of Map<String, String> getMetadata() API in the code happy
    * by internally converting it.
    */
-  private Map<String, String> deserializeMetadata(byte[] bytes)
+  private Map<String, String> deserializeMetadata(ByteBuffer buffer)
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java:343]: Can we add a comment / note stating that the caller of this method assumes that ByteBuffer position is moved accordingly by reads done in this method and so we should use relative get/read methods ?

File: pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java [hunk 2]
```diff
@@ -350,32 +340,29 @@ private byte[] serializeMetadata()
...
+    Map<String, String> metadata = new HashMap<>();
+    for (int i = 0; i < numEntries; i++) {
+      int keyId = buffer.getInt();
+      MetadataKey key = MetadataKey.getByOrdinal(keyId);
+      // Ignore unknown keys.
+      if (key == null) {
+        continue;
+      }
+      if (key.getValueType() == MetadataValueType.INT) {
+        String value = "" + buffer.getInt();
```
Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java:355]: Curious if this is better / faster compared to `String.valueOf()` ?

Comment (siddharthteotia) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datatable/DataTableImplV3.java:355]: Curious if this is better / faster compared to `String.valueOf()` ?

**Review by siddharthteotia — APPROVED** [review]
I feel we should change / enhance the `DataTableSerDeTest` (testDataTableMetadataBytesLayout) unit test to test the ByteBuffer based deserialization although this may probably have enough coverage already via the inter-segment query execution unit tests

**Comment by richardstartin** [comment]
> I feel we should change / enhance the `DataTableSerDeTest` (testDataTableMetadataBytesLayout) unit test to test the ByteBuffer based deserialization or may be SerDe code may probably have enough coverage already via the inter-segment query execution tests ?

OK I will add tests

**Comment by richardstartin** [comment]
Actually these changes are covered by existing tests, nothing could be added without redundancy.

**Pull request merged**